course4/lecture4_en.md

```diff
@@ -0,0 +1,306 @@
+# Modern Programming Concepts: Tuples, Structures, and Enumerated Types
+
+Hello everyone, welcome to the open course on modern programming concepts brought to you by the IDEA's Digital Infrastructure Innovation. Today's topic is tuples, structures, and enumerated types. In a nutshell, we will be exploring how to customize data structures in MoonBit.
+
+# Review: Tuples
+
+First, review the fundamental data type in MoonBit introduced in course 2: Tuples.
+
+A MoonBit tuple is a composite of data which have different types with a fixed length. In comparison, lists are collections of data which have the same type with arbitrary lengths. For example, the length of the list below is not fixed, but the values stored must all be of the character type. In the previous lesson, we didn't discuss why `Cons` is named `Cons`. It is an abbreviation for `Construct`.
+
+```moonbit
+Cons('H', Cons('i', Cons('!', Nil)))
+```
+
+The definition of tuples is represented by expressions enclosed in parentheses and separated by commas. The types are also specified using the same syntax, as in the definition of personal identity information here: `("Bob", 2023, 10, 24): (String, Int, Int, Int)`.
+
+Tuple members are accessed by indexes, starting from 0 on the left. For example, `(2023, 10, 24).0 == 2023`.
+
+# Cartesian Product
+
+You may be familiar with the concept of the Cartesian product. The Cartesian product of two sets is a set where all elements are ordered pairs formed by the elements of the original two sets. For example, the Cartesian product of the set of card suits and the numbers 1 to 52 forms all possible ordered pairs of suits and numbers. 
+
+{ ♥️ ♦️ ♠️ ♣️ } $\times \{ n \in \mathbb{N} | 1 \leq n \leq 52 \}$ 
+
+Tuples, on the other hand, go beyond the Cartesian product of two sets; they represent the Cartesian product of multiple sets, making them more accurately termed as ordered sets. Consequently, tuples are also known as product types. You might wonder if there are sum types alongside product types. We will introduce sum types and explore the concepts of **zero** and **one** later.
```

Comment:
I didn't know ordered sets means cartesian product of multiple sets. May I have a resource please?

Reply:
My mistake here. The ordered set is not the idea we need here. Maybe the following is more appropriate: 

Beyond the Cartesian product of two sets tuples represent the Cartesian product of multiple sets, named n-tuple. Consequently, tuples are also known as product types. You might wonder if there are sum types alongside product types. We will introduce sum types and explore the concepts of **zero** and **one** later.

Reply:
I think I now understand what you meant by ordered sets. n-tuple may be better.